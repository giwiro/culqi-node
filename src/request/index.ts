import https from 'https';
import querystring from 'querystring';
import {RequestOptions, IncomingMessage} from 'http';
import vars from '../vars';

export type RequestMethod = 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE';

export type HttpProvider = typeof https;

export type HttpRequestOptions = {
  method?: RequestMethod;
  useSecureEndpoint?: true;
  headers?: Record<string, string>;
  queryParams?: Record<string, string>;
  body?: Record<string, unknown>;
  // This is for testing purposes only
  _httpProvider?: HttpProvider;
};

function _httpResponseCallback<T>(
  res: IncomingMessage,
  resolve: (r: T) => void,
  reject: (e: Error | object) => void
) {
  let buffer = '';
  const statusCode = res.statusCode;

  res.on('data', (chunk) => (buffer += chunk));

  res.on('end', () => {
    let r;
    if (buffer && buffer.length) {
      try {
        r = JSON.parse(buffer);
      } catch (e) {
        return reject(
          new Error(`Response was not a valid JSON. Response: ${buffer}`)
        );
      }
    }
    if (typeof statusCode !== 'number')
      return reject(new Error('Status code is undefined'));

    if (Math.round(statusCode / 100) !== 2) return reject(r);

    return resolve(r);
  });
}

function request<T>(path: string, options?: HttpRequestOptions): Promise<T> {
  const body = options && options.body ? JSON.stringify(options.body) : '';
  const httpProvider =
    options && options._httpProvider ? options._httpProvider : https;

  const defaultHeaders = {
    'Content-Type': 'application/json',
    'Content-Length': body.length,
    'User-Agent': 'culqi-node',
    Authorization: `Bearer ${vars.privateKey}`,
  };

  const opts: RequestOptions = {
    host: vars.baseEndpoint.domain,
    path: `${vars.baseEndpoint.basePath}${path}`,
    headers: {...defaultHeaders},
    method: 'GET',
  };

  if (options) {
    if (options.useSecureEndpoint) {
      if (!vars.publicKey) throw new Error("Provide 'publicKey' property");
      opts.host = vars.baseSecureEndpoint.domain;
      opts.path = `${vars.baseSecureEndpoint.basePath}${path}`;
      opts.headers = {
        ...opts.headers,
        Authorization: `Bearer ${vars.publicKey}`,
      };
    }
    if (options.headers) opts.headers = {...opts.headers, ...options.headers};
    if (options.method) opts.method = options.method;
    if (options.queryParams && Object.keys(options.queryParams).length > 0)
      opts.path = `${opts.path}?${querystring.stringify(options.queryParams)}`;
  }

  return new Promise<T>((resolve, reject) => {
    const req = httpProvider.request(opts, (res: IncomingMessage) => {
      _httpResponseCallback<T>(res, resolve, reject);
    });
    if (body) req.write(body);
    req.end();
  });
}

function get<T>(
  path: string,
  queryParams?: Record<string, string>,
  options?: Partial<HttpRequestOptions>
): Promise<T> {
  return request<T>(path, {...options, queryParams, method: 'GET'});
}

function post<T>(
  path: string,
  body?: Record<string, unknown>,
  options?: Partial<HttpRequestOptions>
): Promise<T> {
  return request<T>(path, {...options, body, method: 'POST'});
}

function delete_<T>(
  path: string,
  options?: Partial<HttpRequestOptions>
): Promise<T> {
  return request<T>(path, {...options, method: 'DELETE'});
}

function patch<T>(
  path: string,
  body?: Record<string, unknown>,
  options?: Partial<HttpRequestOptions>
): Promise<T> {
  return request<T>(path, {...options, body, method: 'PATCH'});
}

export {get, post, patch, delete_, request, _httpResponseCallback};
