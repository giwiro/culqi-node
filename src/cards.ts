import {get, post, patch, delete_, HttpRequestOptions} from './request';
import vars from './vars';

export type Card = {
  object: string;
  id: string;
  type: string;
  email: string;
  creation_date: number;
  card_number: string;
  last_four: string;
  active: boolean;
  iin: {
    object: string;
    bin: string;
    card_brand: string;
    card_type: string;
    card_category: string;
    issuer?: {
      name: string;
      country: string;
      country_code: string;
      website: string;
      phone_number: string;
    };
    installments_allowed: number[];
  };
  client: {
    ip: string;
    ip_country: string;
    ip_country_code: string;
    browser: string;
    device_fingerprint: string;
    device_type: string;
  };
  metadata: Record<string, string>;
};

export type CreateCardRequest = {
  customer_id: string;
  token_id: string;
};

export type GetCardRequest = {
  id: string;
};

export type GetCardsRequest = {
  creation_date?: string;
  creation_date_to?: string;
  creation_date_from?: string;
  card_brand?: string;
  card_type?: string;
  device_type?: string;
  bin?: string;
  country_code?: string;
  limit?: string;
  before?: string;
  after?: string;
};

export type GetCardsResponse = {
  data: Card[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

export type UpdateCardRequest = {
  id: string;
  metadata?: Record<string, string>;
};

export type DeleteCardRequest = {
  id: string;
};

export type DeleteCardResponse = {
  id: string;
  deleted: boolean;
  merchant_message: string;
};

export const cards = {
  createCard: (
    req: CreateCardRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => post<Card>(vars.basePaths.cards, req, extraHttpOptions),
  getCard: (
    req: GetCardRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Card>(`${vars.basePaths.cards}/${req.id}`, undefined, extraHttpOptions),
  getCards: (
    req?: GetCardsRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetCardsResponse>(
      vars.basePaths.cards,
      req as Record<string, string>,
      extraHttpOptions
    ),
  updateCard: (
    req: UpdateCardRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => patch<Card>(`${vars.basePaths.cards}/${req.id}`, req, extraHttpOptions),
  deleteCard: (
    req: DeleteCardRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    delete_<DeleteCardResponse>(
      `${vars.basePaths.cards}/${req.id}`,
      extraHttpOptions
    ),
};
