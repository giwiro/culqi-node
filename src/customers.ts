import {get, post, patch, delete_, HttpRequestOptions} from './request';
import vars from './vars';

export type Customer = {
  object: string;
  id: string;
  creation_date: number;
  email: string;
  antifraud_details: {
    country_code: string;
    first_name: string;
    last_name: string;
    address_city: string;
    address: string;
    phone: string;
    object: string;
  };
  metadata: Record<string, string>;
};

export type CreateCustomerRequest = {
  first_name: string;
  last_name: string;
  email: string;
  address: string;
  address_city: string;
  country_code: string;
  phone_number: string;
};

export type GetCustomerRequest = {
  id: string;
};

export type GetCustomersRequest = {
  first_name?: string;
  last_name?: string;
  email?: string;
  address?: string;
  address_city?: string;
  phone_number?: string;
  country_code?: string;
  limit?: string;
  before?: string;
  after?: string;
};

export type GetCustomersResponse = {
  data: Customer[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

export type UpdateCustomerRequest = {
  id: string;
  metadata?: Record<string, string>;
};

export type DeleteCustomerRequest = {
  id: string;
};

export type DeleteCustomerResponse = {
  id: string;
  deleted: boolean;
  merchant_message: string;
};

export const customers = {
  createCustomer: (
    req: CreateCustomerRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => post<Customer>(vars.basePaths.customers, req, extraHttpOptions),
  getCustomer: (
    req: GetCustomerRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Customer>(
      `${vars.basePaths.customers}/${req.id}`,
      undefined,
      extraHttpOptions
    ),
  getCustomers: (
    req?: GetCustomersRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetCustomersResponse>(
      vars.basePaths.customers,
      req as Record<string, string>,
      extraHttpOptions
    ),
  updateCustomer: (
    req: UpdateCustomerRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    patch<Customer>(
      `${vars.basePaths.customers}/${req.id}`,
      req,
      extraHttpOptions
    ),
  deleteCustomer: (
    req: DeleteCustomerRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    delete_<DeleteCustomerResponse>(
      `${vars.basePaths.customers}/${req.id}`,
      extraHttpOptions
    ),
};
