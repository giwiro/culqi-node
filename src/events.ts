import {get, HttpRequestOptions} from './request';
import vars from './vars';

export type Event = {
  object: string;
  type: string;
  data: Record<string, unknown>;
};

export type GetEventRequest = {
  id: string;
};

export type GetEventsRequest = {
  type?: string;
  creation_date?: string;
  creation_date_from?: string;
  creation_date_to?: string;
  limit?: string;
  before?: string;
  after?: string;
};

export type GetEventsResponse = {
  data: Event[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

export const events = {
  getEvent: (
    req: GetEventRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Event>(
      `${vars.basePaths.events}/${req.id}`,
      undefined,
      extraHttpOptions
    ),
  getEvents: (
    req?: GetEventsRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetEventsResponse>(
      vars.basePaths.events,
      req as Record<string, string>,
      extraHttpOptions
    ),
};
