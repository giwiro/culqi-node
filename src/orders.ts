import {get, post, patch, delete_, HttpRequestOptions} from './request';
import vars from './vars';

export type Order = {
  object: string;
  id: string;
  amount: number;
  payment_code: string;
  currency_code: string;
  description: string;
  order_number: string;
  state: string;
  total_fee: number;
  net_amount: number;
  fee_details: string;
  creation_date: number;
  expiration_date: number;
  updated_at: number;
  paid_at: number;
  available_on: null;
  metadata: Record<string, string>;
};

export type CreateOrderRequest = {
  amount: number;
  currency_code: string;
  description: string;
  order_number: string;
  client_details: {
    first_name: string;
    last_name: string;
    email: string;
    phone_number: string;
  };
  expiration_date: number;
};

export type ConfirmOrderRequest = {
  id: string;
};

export type GetOrderRequest = {
  id: string;
};

export type GetOrdersRequest = {
  amount?: string;
  min_amount?: string;
  max_amount?: string;
  creation_date?: string;
  creation_date_from?: string;
  creation_date_to?: string;
  state?: string;
  limit?: string;
  before?: string;
  after?: string;
};

export type GetOrdersResponse = {
  data: Order[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

export type UpdateOrderRequest = {
  id: string;
  expiration_date?: number;
  metadata?: Record<string, string>;
};

export type DeleteOrderRequest = {
  id: string;
};

export type DeleteOrderResponse = {
  id: string;
  deleted: boolean;
  merchant_message: string;
};

export const orders = {
  createOrder: (
    req: CreateOrderRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => post<Order>(vars.basePaths.orders, req, extraHttpOptions),
  confirmOrder: (
    req: ConfirmOrderRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    post<Order>(
      `${vars.basePaths.orders}/${req.id}/confirm`,
      req,
      extraHttpOptions
    ),
  getOrder: (
    req: GetOrderRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Order>(
      `${vars.basePaths.orders}/${req.id}`,
      undefined,
      extraHttpOptions
    ),
  getOrders: (
    req?: GetOrdersRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetOrdersResponse>(
      vars.basePaths.orders,
      req as Record<string, string>,
      extraHttpOptions
    ),
  updateOrder: (
    req: UpdateOrderRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    patch<Order>(`${vars.basePaths.orders}/${req.id}`, req, extraHttpOptions),
  deleteOrder: (
    req: DeleteOrderRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    delete_<DeleteOrderResponse>(
      `${vars.basePaths.orders}/${req.id}`,
      extraHttpOptions
    ),
};
