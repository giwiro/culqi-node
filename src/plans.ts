import {get, post, patch, delete_, HttpRequestOptions} from './request';
import vars from './vars';

export type PlanInterval = 'Dias' | 'Semanas' | 'Meses' | 'AÃ±os';

export type Plan = {
  object: string;
  id: string;
  name: string;
  amount: number;
  currency_code: string;
  interval_count: number;
  interval: PlanInterval;
  limit: number;
  trial_days: number;
  total_subscriptions: number;
  metadata: Record<string, string>;
};

export type CreatePlanRequest = {
  name: string;
  amount: number;
  currency_code: string;
  interval: PlanInterval;
  interval_count: number;
  limit: number;
};

export type GetPlanRequest = {
  id: string;
};

export type GetPlansRequest = {
  amount?: string;
  min_amount?: string;
  max_amount?: string;
  creation_date_from?: string;
  creation_date_to?: string;
  limit?: string;
  before?: string;
  after?: string;
};

export type GetPlansResponse = {
  data: Plan[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

export type UpdatePlanRequest = {
  id: string;
  metadata?: Record<string, string>;
};

export type DeletePlanRequest = {
  id: string;
};

export type DeletePlanResponse = {
  id: string;
  deleted: boolean;
  merchant_message: string;
};

export const plans = {
  createPlan: (
    req: CreatePlanRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => post<Plan>(vars.basePaths.plans, req, extraHttpOptions),
  getPlan: (
    req: GetPlanRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Plan>(`${vars.basePaths.plans}/${req.id}`, undefined, extraHttpOptions),
  getPlans: (
    req?: GetPlansRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetPlansResponse>(
      vars.basePaths.plans,
      req as Record<string, string>,
      extraHttpOptions
    ),
  updatePlan: (
    req: UpdatePlanRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => patch<Plan>(`${vars.basePaths.plans}/${req.id}`, req, extraHttpOptions),
  deletePlan: (
    req: DeletePlanRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    delete_<DeletePlanResponse>(
      `${vars.basePaths.plans}/${req.id}`,
      extraHttpOptions
    ),
};
