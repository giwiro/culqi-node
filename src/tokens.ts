import {get, post, patch, HttpRequestOptions} from './request';
import vars from './vars';

export type Token = {
  object: string;
  id: string;
  type: string;
  email: string;
  creation_date: number;
  card_number: string;
  last_four: string;
  active: boolean;
  iin: {
    object: string;
    bin: string;
    card_brand: string;
    card_type: string;
    card_category: string;
    issuer: {
      name: string;
      country: string;
      country_code: string;
      website: string;
      phone_number: string;
    };
    installments_allowed: number[];
  };
  client: {
    ip: string;
    ip_country: string;
    ip_country_code: string;
    browser: string;
    device_fingerprint: string;
    device_type: string;
  };
  metadata: Record<string, string>;
};

export type CreateTokenRequest = {
  card_number: string;
  cvv: string;
  expiration_month: string;
  expiration_year: string;
  email: string;
  metadata?: Record<string, string>;
};

export type GetTokenRequest = {
  id: string;
};

export type GetTokensRequest = {
  creation_date?: string;
  creation_date_to?: string;
  card_brand?: string;
  card_type?: string;
  device_type?: string;
  bin?: string;
  country_code?: string;
  limit?: string;
  before?: string;
  after?: string;
};

export type GetTokensResponse = {
  data: Token[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

export type UpdateTokenRequest = {
  id: string;
  metadata?: Record<string, string>;
};

export const tokens = {
  createToken: (
    req: CreateTokenRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    post<Token>(vars.basePaths.tokens, req, {
      ...extraHttpOptions,
      useSecureEndpoint: true,
    }),
  getTokens: (
    req?: GetTokensRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetTokensResponse>(
      vars.basePaths.tokens,
      req as Record<string, string>,
      extraHttpOptions
    ),
  getToken: (
    req: GetTokenRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Token>(
      `${vars.basePaths.tokens}/${req.id}`,
      undefined,
      extraHttpOptions
    ),
  updateToken: (
    req: UpdateTokenRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    patch<Token>(`${vars.basePaths.tokens}/${req.id}`, req, extraHttpOptions),
};
