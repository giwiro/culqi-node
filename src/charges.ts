import {get, post, patch, HttpRequestOptions} from './request';
import vars from './vars';
import {Card} from './cards';

export type Charge = {
  object: string;
  id: string;
  creation_date: number;
  amount: number;
  amount_refunded: number;
  current_amount: number;
  installments: number;
  installments_amount: number;
  currency: string;
  email: string;
  description: string;
  source: Card;
  outcome: {
    type: string;
    code: string;
    merchant_message: string;
    user_message: string;
  };
  fraud_score: number;
  antifraud_details: {
    first_name: string;
    last_name: string;
    address: string;
    address_city: string;
    country_code: string;
    phone: string;
    object: string;
  };
  dispute: boolean;
  capture: boolean;
  reference_code: string;
  metadata: Record<string, string>;
  total_fee: number;
  fee_details: {
    fixed_fee: Record<string, string | number>;
    variable_fee: {
      currency_code: string;
      commision: number;
      total: number;
    };
  };
  paid: boolean;
  statement_descriptor: string;
  total_fee_taxes: number;
  transfer_amount: number;
  duplicated: boolean;
};

export type CreateChargeRequest = {
  amount: string;
  currency_code: string;
  email: string;
  source_id: string;
  capture?: boolean;
  description?: string;
  installments?: number;
  metadata?: Record<string, string>;
  antifraud_details?: {
    first_name: string;
    last_name: string;
    address: string;
    address_city: string;
    country_code: string;
    phone: string;
  };
};

export type GetChargeRequest = {
  id: string;
};

export type GetChargesRequest = {
  amount?: string;
  min_amount?: string;
  max_amount?: string;
  installments?: string;
  min_installments?: string;
  max_installments?: string;
  currency_code?: string;
  code?: string;
  decline_code?: string;
  fraud_score?: string;
  min_fraud_score?: string;
  max_fraud_score?: string;
  first_name?: string;
  last_name?: string;
  email?: string;
  address?: string;
  address_city?: string;
  phone_number?: string;
  country_code?: string;
  dispute?: string;
  captured?: string;
  duplicated?: string;
  paid?: string;
  customer_id?: string;
  reference?: string;
  creation_date?: string;
  creation_date_from?: string;
  creation_date_to?: string;
  fee?: string;
  min_fee?: string;
  max_fee?: string;
  card_brand?: string;
  card_type?: string;
  device_type?: string;
  bin?: string;
  limit?: string;
  before?: string;
  after?: string;
};

export type GetChargesResponse = {
  data: Charge[];
  paging: {
    previous: string;
    next: string;
    cursors: {
      before: string;
      after: string;
    };
    remaining_items: number;
  };
};

export type UpdateChargeRequest = {
  id: string;
  metadata?: Record<string, string>;
};

export type CaptureChargeRequest = {
  id: string;
};

export const charges = {
  createCharge: (
    req: CreateChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) => post<Charge>(vars.basePaths.charges, req, extraHttpOptions),
  getCharge: (
    req: GetChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<Charge>(
      `${vars.basePaths.charges}/${req.id}`,
      undefined,
      extraHttpOptions
    ),
  getCharges: (
    req?: GetChargesRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    get<GetChargesResponse>(
      vars.basePaths.charges,
      req as Record<string, string>,
      extraHttpOptions
    ),
  updateCharge: (
    req: UpdateChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    patch<Charge>(`${vars.basePaths.charges}/${req.id}`, req, extraHttpOptions),
  captureCharge: (
    req: CaptureChargeRequest,
    extraHttpOptions?: Partial<HttpRequestOptions>
  ) =>
    post<Charge>(
      `${vars.basePaths.charges}/${req.id}/capture`,
      req,
      extraHttpOptions
    ),
};
