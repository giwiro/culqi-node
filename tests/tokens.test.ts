import {tokens} from '../src/tokens';
import vars from '../src/vars';
import {httpMockFactory} from './utils/request';
import {generateCreateTokenRequest} from './utils/card';
import {HttpProvider} from '../src/request';
import {sleep} from './utils/time';

describe('tokens', () => {
  let publicKey: string;
  let privateKey: string;

  let createdTokenId: string;

  beforeEach(() => {
    privateKey = process.env.CULQI_PRIVATE_KEY || '';
    publicKey = process.env.CULQI_PUBLIC_KEY || '';
    vars.privateKey = privateKey;
    vars.publicKey = publicKey;
    jest.clearAllMocks();
  });

  describe('createToken', () => {
    it('should change path', () => {
      const mockedHttps = httpMockFactory();
      tokens.createToken(generateCreateTokenRequest(), {
        _httpProvider: mockedHttps as unknown as HttpProvider,
      });
      const c = mockedHttps.request.mock.calls[0][0];
      expect(c.path).toMatchSnapshot();
    });

    it('should create token', async () => {
      await sleep(5000);
      const resp = await tokens.createToken(generateCreateTokenRequest());
      createdTokenId = resp.id;
      expect(resp.object).toMatchSnapshot();
      expect(resp.type).toMatchSnapshot();
    });
  });

  describe('getTokens', () => {
    it('should change path', () => {
      const mockedHttps = httpMockFactory();
      tokens.getTokens(undefined, {
        _httpProvider: mockedHttps as unknown as HttpProvider,
      });
      const c = mockedHttps.request.mock.calls[0][0];
      expect(c.path).toMatchSnapshot();
    });

    it('should get all tokens', async () => {
      const resp = await tokens.getTokens();
      expect(resp.data.length).toBeGreaterThan(0);
    });
  });

  describe('getToken', () => {
    it('should change path', () => {
      const mockedHttps = httpMockFactory();
      tokens.getToken(
        {
          id: createdTokenId,
        },
        {
          _httpProvider: mockedHttps as unknown as HttpProvider,
        }
      );
      const c = mockedHttps.request.mock.calls[0][0];
      expect(c.path).toMatch(
        `${vars.baseEndpoint.basePath}${vars.basePaths.tokens}/${createdTokenId}`
      );
    });

    it('should get token', async () => {
      const resp = await tokens.getToken({
        id: createdTokenId,
      });
      expect(resp.object).toMatchSnapshot();
      expect(resp.type).toMatchSnapshot();
    });
  });

  describe('updateToken', () => {
    it('should change path', () => {
      const mockedHttps = httpMockFactory();
      tokens.updateToken(
        {
          id: createdTokenId,
          metadata: {
            foo: 'bar',
          },
        },
        {
          _httpProvider: mockedHttps as unknown as HttpProvider,
        }
      );
      const c = mockedHttps.request.mock.calls[0][0];
      expect(c.path).toMatch(
        `${vars.baseEndpoint.basePath}${vars.basePaths.tokens}/${createdTokenId}`
      );
    });

    // 500 in Culqi dev
    // eslint-disable-next-line jest/no-commented-out-tests
    /*it('should update token', async () => {
      const resp = await tokens.updateToken({
        id: createdTokenId,
        metadata: {
          foo: 'bar',
        },
      });
      expect(resp.object).toMatchSnapshot();
      expect(resp.type).toMatchSnapshot();
      expect(resp.metadata).toEqual({
        foo: 'bar',
      });
    });*/
  });
});
