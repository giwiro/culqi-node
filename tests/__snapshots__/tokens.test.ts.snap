// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokens createToken should change path 1`] = `"/v2/tokens"`;

exports[`tokens createToken should create token 1`] = `"token"`;

exports[`tokens createToken should create token 2`] = `"card"`;

exports[`tokens getToken should get token 1`] = `"token"`;

exports[`tokens getToken should get token 2`] = `"card"`;

exports[`tokens getTokens should change path 1`] = `"/v2/tokens"`;
